src/main/java/com/mongodb/spark/sql/connector/read/MongoStreamPartitionReader.java

```diff
@@ -0,0 +1,204 @@
+ *
+ * <p>It's responsible for outputting data for a continuous stream using change streams.
+ */
+public class MongoStreamPartitionReader implements ContinuousPartitionReader<InternalRow> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongoPartitionReader.class);
+  private static final String INVALIDATE = "invalidate";
+  private static final ResumeTokenPartitionOffset EMPTY_RESUME_TOKEN_PARTITION_OFFSET =
+      new ResumeTokenPartitionOffset(new BsonDocument());
+  private final MongoInputPartition partition;
+  private final BsonDocumentToRowConverter bsonDocumentToRowConverter;
+  private final ReadConfig readConfig;
+
+  private ResumeTokenPartitionOffset lastOffset;
+  private InternalRow currentRow;
+  private boolean closed = false;
+  private MongoClient mongoClient;
+  private MongoChangeStreamCursor<BsonDocument> changeStreamCursor;
+
+  /**
+   * Construct a new instance
+   *
+   * @param partition the partition
+   * @param bsonDocumentToRowConverter the converter from {@link BsonDocument} to {@link
+   *     InternalRow}
+   * @param readConfig the read configuration for reading from the partition
+   */
+  public MongoStreamPartitionReader(
+      final MongoInputPartition partition,
+      final BsonDocumentToRowConverter bsonDocumentToRowConverter,
+      final ReadConfig readConfig) {
+    this.partition = partition;
+    this.bsonDocumentToRowConverter = bsonDocumentToRowConverter;
+
+    this.readConfig = readConfig;
+    this.currentRow = null;
+    this.lastOffset =
+        new ResumeTokenPartitionOffset(partition.getResumeTokenOffset().getResumeToken());
+
+    LOGGER.debug(
+        "Creating partition reader for: PartitionId: {} with Schema: {}",
+        partition.getPartitionId(),
+        bsonDocumentToRowConverter.getSchema());
+  }
+
+  @Override
+  public PartitionOffset getOffset() {
+    Assertions.ensureState(
+        () -> !closed, () -> "Cannot call getOffset() on a closed PartitionReader.");
+    LOGGER.trace("getOffset called, returning: {}", lastOffset);
+    return lastOffset;
+  }
+
+  @Override
+  public boolean next() {
+    Assertions.ensureState(() -> !closed, () -> "Cannot call next() on a closed PartitionReader.");
+    return withCursor(
+        c -> {
+          try {
+            BsonDocument next = c.next();
+            lastOffset = new ResumeTokenPartitionOffset(c.getResumeToken());
+            if (next.getString("operationType", new BsonString(""))
+                .getValue()
+                .toLowerCase(Locale.ROOT)
+                .equals(INVALIDATE)) {
+              LOGGER.info(
+                  "Change stream cursor has been invalidated. This happens when a collection is dropped. Closing cursor.");
+              lastOffset = EMPTY_RESUME_TOKEN_PARTITION_OFFSET;
+              releaseCursor();
+              releaseClient();
+            }
+            currentRow = bsonDocumentToRowConverter.toInternalRow(next);
```

Comment:
Instead of reading the actual documents here, we are reading [change events](https://docs.mongodb.com/manual/reference/change-events/#std-label-change-stream-output), which is different from what `MongoPartitionReader` reads. This, in turn, means that `MongoScan.toContinuousStream` and `MongoScan.toBatch` read change events vs actual documents. Is it not the case that they are supposed to read the same data but in different ways?

Reply:
Both `MongoPartitionReader` and `MongoStreamPartitionReader` read bson documents as change events are just bson documents. So both partitioners read bson documents and convert them to output InternalRow data.

Reply:
Sure, they both read `BsonDocument`. The difference I am talking about is in what this document is: in case of `MongoPartitionReader` the document is what an aggregation pipeline outputs, while in case of `MongoStreamPartitionReader` the document is a change event. My expectation would be that `Scan.toBatch` and `Scan.toContinuousStream` allow me to read the same documents, and that's not the case in Mongo Spark. Is my expectation groundless and users don't expect this?

Either way, it seems worth at the very least documenting (including javadocs) that `MongoScan.toBatch` and `MongoScan.toContinuousStream` read different documents: the former reads documents outputted by an aggregation pipeline (most likely simply documents stored in a collection), while the latter reads change events.

Reply:
Well to be devils advocate change events are just aggregation results also ;) but I get your point - will clarify.

Will clarify the java doc - but the real place this needs documenting (as users wont see the java doc) is in the usage documentation.  Also there will be an option to just output full documents only in another ticket - but change events are more than just insert events.